src/host-config/redsocks.ts

```diff
@@ -0,0 +1,352 @@
+				continue;
+			}
+
+			const label = blockLabel as keyof RedsocksConfig;
+
+			// Process raw block string to key-value pairs
+			const parsed = RedsocksConf.parseBlock(block, label);
+			if (Object.keys(parsed).length === 0) {
+				log.warn(
+					`Ignoring invalid redsocks.conf configuration block: "${label}"`,
+				);
+				continue;
+			}
+			// Decode to block-specific type
+			if (label === 'base') {
+				const baseDecoded = BaseConfig.decode(parsed);
+				if (isRight(baseDecoded)) {
+					conf.base = baseDecoded.right;
+				}
+			} else if (label === 'redsocks') {
+				const proxyDecoded = ReadableProxyConfig.decode(parsed);
+				if (isRight(proxyDecoded)) {
+					conf.redsocks = proxyDecoded.right;
+				} else {
+					log.warn(
+						[
+							'Unexpected redsocks.conf configuration detected, proxy may not behave as expected:',
+						]
+							.concat(Reporter.report(proxyDecoded))
+							.map(maskAuthData)
+							.join('\n'),
+					);
+					// if parsed has all keys, allow it to be used for backwards compatibility
+					if (Object.keys(parsed).length === 7) {
+						conf.redsocks = parsed as unknown as ReadableProxyConfig;
+					}
+				}
+			}
+		}
+
+		return conf;
+	}
+
+	public static patch(
+		currentConf: RedsocksConfig,
+		inputConf: Partial<{
+			base: BaseConfig;
+			redsocks: Partial<WriteableProxyConfig>;
+		}>,
+	): RedsocksConfig {
+		const patchedConf: RedsocksConfig = {};
+
+		// If input is empty, redsocks config should be removed
+		if (!inputConf || Object.keys(inputConf).length === 0) {
+			return patchedConf;
+		}
+
+		if (
+			inputConf.redsocks &&
+			typeof inputConf.redsocks === 'object' &&
+			Object.keys(inputConf.redsocks).length > 0
+		) {
+			// Only include base config if redsocks config is present
+			patchedConf.base = {
+				...defaultBaseConfig,
+			};
+			// This method assumes that currentConf is a full ProxyConfig
+			// for backwards compatibility, so patchedConf.redsocks should
+			// also be a full ProxyConfig.
+			patchedConf.redsocks = {
+				...currentConf.redsocks,
+				...inputConf.redsocks,
+				// local_ip & local_port are not configurable
+				local_ip: '127.0.0.1',
+				local_port: 12345,
+			} as ReadableProxyConfig;
+		}
+
+		return patchedConf;
+	}
+
+	private static stringifyBlock(
+		label: string,
+		block: BaseConfig | ReadableProxyConfig,
+	): string {
+		const lines = Object.entries(block).map(
+			([key, value]) => `\t${key} = ${value};`,
+		);
+		return `${label} {\n${lines.join('\n')}\n}`;
+	}
+
+	private static parseBlock(
+		block: string,
+		label: string,
+	): Record<string, string> {
+		const parsedBlock: Record<string, string> = {};
+		// Skip first and last line (`$BLOCK_LABEL {` and `}`)
+		const lines = block.split('\n').slice(1, -1);
+		for (const line of lines) {
+			let [key, value] = line.split('=').map((s) => s.trim());
+
+			// Remove trailing semicolon
+			if (value) {
+				if (value.endsWith(';')) {
```

Comment:
[Updated commit](https://github.com/balena-os/balena-supervisor/pull/2329/commits/0ce259c1845c9137b07618aa6aafa54361771693#diff-87da4178ce07f20bae291890a95e15b8bf7990d89aeb29ec6bfe1f77b508c3e5R76) splits by newline and optionally semicolon, there's no semicolon validation otherwise now